### PR TITLE
Custom dimension tile wrapper

### PR DESCRIPTION
**Pull request by nfmohit**
## Summary

Addresses issue:

- #7601

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/components/KeyMetrics/MetricTileWrapper.js
Comment (tofumatt) [inline assets/js/components/KeyMetrics/MetricTileWrapper.js:262]: No need to have the separate translation string here, this should work:

```suggestion
								{ createInterpolateElement(
									__(
										'Permissions updated? <a>Retry</a>',
										'google-site-kit'
									),
									{
										a: (
											<Link
												onClick={
													handleCreateCustomDimensions
												}
											/>
										),
									}
								) }
```

File: assets/js/components/KeyMetrics/MetricTileWrapper.js
Comment (tofumatt) [inline assets/js/components/KeyMetrics/MetricTileWrapper.js:280]: Same here:

```suggestion
								{ createInterpolateElement(
									__(
										'You’ll need to contact your administrator. <a>Learn more</a>',
										'google-site-kit'
									),
									{
										a: (
											<Link href={ helpLink } external />
										),
									}
								) }
```

File: assets/js/components/KeyMetrics/MetricTileWrapper.js
Comment (tofumatt) [inline assets/js/components/KeyMetrics/MetricTileWrapper.js:314]: Same here:

```suggestion
							{ createInterpolateElement(
								__(
									'Retry didn’t work? <a>Learn more</a>',
									'google-site-kit'
								),
								{
									a: (
										<Link href={ helpLink } external />
									),
								}
							) }
```

File: assets/js/components/KeyMetrics/MetricTileWrapper.js
Comment (tofumatt) [reply assets/js/components/KeyMetrics/MetricTileWrapper.js:262]: (I realise an argument to separating out the small translation strings is that things like `__( 'Retry' )` and `__( 'Learn more' )` are already translated, but keeping as much context in one string is usually best, and since the main string needs to be translated anyway, I think it's fine 😄 )

**Review by tofumatt — DISMISSED** [review]
Looks great!

A few notes on the translation strings here, but they're minor.

------

I did encounter an issue when following the QA Brief:

> Upon saving your selection, you should be directed to the OAuth flow. Cancel the OAuth flow to return to the SK dashboard.

I wasn't redirected, instead I was immediately shown this screen:

CleanShot 2023-10-13 at 15 53 40

> Verify that clicking on "Update" makes the tiles go on a loading state

I don't see the loading state if I haven't granted OAuth access and am being redirected to the OAuth flow:

https://github.com/google/site-kit-wp/assets/90871/648230a9-1000-4c0b-8413-b7dcbd2e0725

Is that expected?

Even after I've granted access, I don't see a network request when clicking "Update":

https://github.com/google/site-kit-wp/assets/90871/8e49950c-3430-4264-8ba6-5924674da3eb

File: assets/js/components/KeyMetrics/MetricTileWrapper.js
Comment (nfmohit) [reply assets/js/components/KeyMetrics/MetricTileWrapper.js]: Updated, thank you!

File: assets/js/components/KeyMetrics/MetricTileWrapper.js
Comment (nfmohit) [reply assets/js/components/KeyMetrics/MetricTileWrapper.js]: Updated, thank you!

File: assets/js/components/KeyMetrics/MetricTileWrapper.js
Comment (nfmohit) [reply assets/js/components/KeyMetrics/MetricTileWrapper.js]: Updated, thank you!

**Comment by nfmohit** [comment]
> A few notes on the translation strings here, but they're minor.

Updated, thank you!

> > Upon saving your selection, you should be directed to the OAuth flow. Cancel the OAuth flow to return to the SK dashboard.
> 
> I wasn't redirected, instead I was immediately shown this screen:

This behaviour is actually dependent on #7599, I think we should probably wait for that to be merged.

> > Verify that clicking on "Update" makes the tiles go on a loading state
> 
> I don't see the loading state if I haven't granted OAuth access and am being redirected to the OAuth flow:
> 
> Is that expected?

I've updated the PR to fix this, thanks!

> Even after I've granted access, I don't see a network request when clicking "Update":

This is also dependent on #7599.

Thank you @tofumatt !

**Comment by tofumatt** [comment]
Cool, I don't think there's anything else for me to review here and @eugene-manuilov took this one first so I'll leave it with him to handle the follow-up here so we don't duplicate effort 🙂

File: assets/js/modules/analytics-4/components/widgets/PopularAuthorsWidget.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/components/widgets/PopularAuthorsWidget.js:36]: This wrapper should be implemented as HOC because we need to make sure that wrapped widgets don't generate report requests when custom dimensions are not created yet. See the `whenActive` HOC that also uses hooks internally: https://github.com/google/site-kit-wp/blob/019635e605e68078aff6fc6f7c1d224a75e1349a/assets/js/util/when-active.js#L49-L53

File: assets/js/components/KeyMetrics/MetricTileLoading.js
Comment (eugene-manuilov) [inline assets/js/components/KeyMetrics/MetricTileLoading.js:1]: This component is needed only because you need to replicate the same layout in the custom dimensions wrapper. This is not needed because we need to use the `MetricTileWrapper` component with the `loading={ true }` property. Then this component can be removed.

File: assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js:97]: How about this:
```suggestion
	const hasCustomDimensions = useSelect( ( select ) => 
		! customDimensions ) || 
		select( MODULES_ANALYTICS_4 ).hasCustomDimensions( customDimensions )
	);
```

File: assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js:106]: Same here:
```suggestion
	const isCreatingCustomDimensions = useSelect( ( select ) =>
		!! customDimensions &&
		customDimensions.some( select( MODULES_ANALYTICS_4 ).isCreatingCustomDimension )
	);
```

File: assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js:123]: We call the same selector twice per each item. This is inefficient. Let's convert this into classical for loop.

File: assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js:191]: If I remember correctly, this will show the insufficient permissions error popup and user will need to manually click on the button to proceed granting required permissions. This is not what we wanted to do when the user clicks on the retry button. AC says: `Clicking on Retry should attempt to create the custom dimensions again.`. So instead of showing the permissions dialog, we need to attempt re-creating custom dimensions again.

File: assets/js/modules/analytics-4/datastore/custom-dimensions.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/datastore/custom-dimensions.js:427]: Hi @eugene-manuilov.

I actually had this same thought when @aaemnnosttv and I were discussing this approach internally. This was Evan's concern:

> I think `isSyncingAvailableCustomDimensions` may be necessary to avoid a race condition, but we can check during implementation. Basically, my concern is that if each component checks `isFetchingSyncAvailableCustomDimensions` before making the request, they'd all see that it is not being fetched and then invoke the fetch simultaneously. If we lock it with a simple state before the fetch which is set synchronously, I think that will be avoided.

Even though I wasn't able to replicate this race condition in my short span of tests, however, I think Evan's concern makes sense and a race condition might occur if we go down that path.

What do you think?

File: assets/js/modules/analytics-4/datastore/custom-dimensions.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/datastore/custom-dimensions.js:280]: Hi @eugene-manuilov.

I actually only added this action for easier datastore mocking in stories and tests (for now). It is not being used elsewhere.

Without an action like this, we'll have to export `possibleCustomDimensions` (as a constant, for example) so that we can use it with `receiveError`.

File: assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js:191]: Hi @eugene-manuilov. We're actually using the `skipModal` property here so that it doesn't show the modal and directly redirects to the OAuth flow. Let me know if that looks good, thanks!

Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js:191]: Ok, I missed that property. Thanks for letting me know. Anyway, you use this callback for all retry buttons, but we don't need to retry permissions for non-insufficient permissions errors. For other errors, we need to just retry the creation action.

File: assets/js/components/KeyMetrics/MetricTileHeader.js
Comment (nfmohit) [reply assets/js/components/KeyMetrics/MetricTileHeader.js:33]: Updated, thanks!

File: assets/js/modules/analytics-4/datastore/custom-dimensions.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/datastore/custom-dimensions.js:255]: Refactored, thanks!

File: assets/js/modules/analytics-4/components/widgets/PopularAuthorsWidget.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/components/widgets/PopularAuthorsWidget.js:36]: Implemented as HOC, please let me know how it looks now, thanks!

File: assets/js/components/KeyMetrics/MetricTileLoading.js
Comment (nfmohit) [reply assets/js/components/KeyMetrics/MetricTileLoading.js:1]: Good point, updated, thanks!

File: assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js:97]: Updated, thanks!

File: assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js]: Updated, thanks!

File: assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js]: Updated, does this look better now? Thanks!

File: assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js]: > Anyway, you use this callback for all retry buttons, but we don't need to retry permissions for non-insufficient permissions errors.

The permission error in this context is a little different from the permission scope error we're dispatching. The permission error here would occur if the user doesn't have access to the connected Analytics property with at least an "editor" role. However, this user will still be able to grant the necessary scope.

This is why I've dispatched the `setPermissionScopeError` for all cases as the Analytics "edit" scope is a prerequisite for creating custom dimensions.

Let me know if that sounds good, thank you!

File: assets/js/modules/analytics-4/utils/withCustomDimensions.test.js
Comment (aaemnnosttv) [inline assets/js/modules/analytics-4/utils/withCustomDimensions.test.js:37]: I find it cleaner to initialize the test `registry` in a `beforeEach` and then pass it to in the render options rather than doing the other way around (we're too flexible! 😄)

This way we can't apply common setup (the way `beforeEach` is intended to be used).

In this case it might only be 
```js
provideUserAuthentication( registry );
provideUserCapabilities( registry );
```

but that does help keep each test a bit more concise.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @nfmohit! This is a big one and looking good. From my POV this looks quite close. There are some things that could be improved but we need to balance that with unblocking other important issues that are waiting for this to land, so let's focus on the most important things at this point and revisit any secondary points in a follow-up. Nice work!

File: assets/js/components/KeyMetrics/MetricTileError.js
Comment (nfmohit) [reply assets/js/components/KeyMetrics/MetricTileError.js]: Excellent catch, updated, thank you!

File: assets/js/components/notifications/CTA.js
Comment (nfmohit) [reply assets/js/components/notifications/CTA.js:106]: Updated, thank you!

File: assets/js/modules/analytics-4/datastore/custom-dimensions.test.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/datastore/custom-dimensions.test.js]: Understood, thank you @aaemnnosttv. I only used the selector here because it was a little inconvenient to access the error in the store because we need to generate the error key using the custom dimension options, which we don't have access to as it is not exported from `custom-dimensions.js`.

However, I've worked around that by adding a local copy of the custom dimension options in the test case and generating the error copy there.

Let me know what you think, thanks!

File: assets/js/modules/analytics-4/utils/withCustomDimensions.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/utils/withCustomDimensions.js:59]: I agree. I'll open a follow-up issue as soon as the PR is approved otherwise, thanks!

File: assets/js/modules/analytics-4/utils/withCustomDimensions.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/utils/withCustomDimensions.js]: Removed, thanks!

File: assets/js/modules/analytics-4/utils/withCustomDimensions.test.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/utils/withCustomDimensions.test.js]: Updated, thank you!

**Comment by nfmohit** [comment]
> Thanks @nfmohit! This is a big one and looking good. From my POV this looks quite close. There are some things that could be improved but we need to balance that with unblocking other important issues that are waiting for this to land, so let's focus on the most important things at this point and revisit any secondary points in a follow-up. Nice work!

Thank you for the kind words and review, @aaemnnosttv. I've addressed the feedback, thanks!

File: assets/js/modules/analytics-4/datastore/custom-dimensions.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/datastore/custom-dimensions.js:243]: No need to pass the propertyID parameter anymore.

File: assets/js/modules/analytics-4/datastore/custom-dimensions.js
Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/datastore/custom-dimensions.js:415]: How is `isSyncingAvailableCustomDimensions` different from `isFetchingSyncAvailableCustomDimensions`? These are completely similar selectors with exactly the same logic. If we worry that isFetching may cause race conditions, then isSyncing has the same problem. I think that if we truly want to be safe here, we should use debounce the sync action and call it in the create dimensions action.

cc @aaemnnosttv

File: assets/js/modules/analytics-4/datastore/custom-dimensions.js
Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/datastore/custom-dimensions.js:268]: Let's make this action to be a utility function then and have the same approach as we use for `provide*` helpers. So we will have something like this: `provideCustomDimensionsError( registry, { customDimension, error } )`.

> Without an action like this, we'll have to export `possibleCustomDimensions` (as a constant, for example)

Fine with me.

File: assets/js/modules/analytics-4/datastore/custom-dimensions.js
Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/datastore/custom-dimensions.js:229]: Replied there. I think we can drop this action.

File: assets/js/modules/analytics-4/utils/withCustomDimensions.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/utils/withCustomDimensions.js:151]: We can use the same shorthands here to keep this component a bit more concise:

```suggestion
			const hasAnalyticsEditScope = useSelect( ( select ) =>
				customDimensions &&
				select( CORE_USER ).hasScope( ANALYTICS_EDIT_SCOPE )
			);
			const isSyncingAvailableCustomDimensions = useSelect(
				( select ) =>
					customDimensions &&
					select( MODULES_ANALYTICS_4 ).isSyncingAvailableCustomDimensions()
			);
```

Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/utils/withCustomDimensions.js:151]: The same applies to all selectors in this file. This will let us have less lines in this component.

File: assets/js/modules/analytics-4/utils/withCustomDimensions.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/utils/withCustomDimensions.js:323]: The `handleCreateCustomDimensions` callback is used specifically for the insufficient permissions error. How will it help if we use it as a handler for the onclick event for this button? When user clicks on this button, we need to retry the failed request, don't we?

File: assets/js/modules/analytics-4/utils/withCustomDimensions.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/utils/withCustomDimensions.js:356]: The same here. How does `handleCreateCustomDimensions` help with updating custom dimensions? It will redirect users to the permissions granting scope. In my understanding we should just reset resolution for the `getReport` selector, right?

File: assets/js/modules/analytics-4/utils/withCustomDimensions.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/utils/withCustomDimensions.js:372]: We also need to give this HOC a name in the same way we do it for `WhenActive`: https://github.com/google/site-kit-wp/blob/8aff60e144bf845cb787ce23f79c96b15a6206cd/assets/js/util/when-active.js#L90-L97

File: assets/js/modules/analytics-4/utils/withCustomDimensions.js
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/utils/withCustomDimensions.js:56]: How are we going to pass report options into this HOC? Usually all our report options are dynamic and depend on the selected date range which means that we can't just pass an object with parameters into this HOC.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, Nahid. Left more feedback for you. Please, let me know what you think.

File: assets/js/modules/analytics-4/components/widgets/PopularAuthorsWidget.stories.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/components/widgets/PopularAuthorsWidget.stories.js]: Removed, thanks!

File: assets/js/modules/analytics-4/datastore/custom-dimensions.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/datastore/custom-dimensions.js]: Updated, thanks!

File: assets/js/modules/analytics-4/utils/withCustomDimensions.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/utils/withCustomDimensions.js]: Updated, thanks!

File: assets/js/modules/analytics-4/utils/withCustomDimensions.js
Comment (nfmohit) [reply assets/js/modules/analytics-4/utils/withCustomDimensions.js:372]: Updated, thanks!

File: assets/js/components/KeyMetrics/MetricTileError.js
Comment (nfmohit) [reply assets/js/components/KeyMetrics/MetricTileError.js:34]: Added back.

**Review by eugene-manuilov — APPROVED** [review]
Ok, looks good to me now. Thanks, @nfmohit.

**Review by tofumatt — COMMENTED** [review]
Looks good to me, but I noticed all E2E tests and the JS tests were failing. I re-ran them a few times and it seems something's actually off with them, can you look before we merge @eugene-manuilov?

**Pull request merged**
